src/List.mo

```diff
@@ -164,15 +164,17 @@ module {
     Result.mapOk(go(xs, null), func (xs : List<R>) : List<R> = reverse(xs))
   };
 
+  /// Append the elements from the reverse of one list to another list.
+  func revAppend<T>(l : List<T>, m : List<T>) : List<T> {
+    switch l {
+      case null { m };
+      case (?(h, t)) { revAppend(t, ?(h, m)) };
+    }
+  };
+
   /// Append the elements from one list to another list.
   public func append<T>(l : List<T>, m : List<T>) : List<T> {
-    func rec(l : List<T>) : List<T> {
-      switch l {
-        case null { m };
-        case (?(h, t)) {?(h,rec(t))};
-      }
-    };
-    rec(l)
+    revAppend(reverse(l), m);
```

Comment:
You have sufficient trust in `tailrec.ml` that you want to write this recursively, instead of just directly write the loop that you want?

Reply:
Would be worth inspecting the generated code to be sure.

Reply:
It would be neat to have a `/// @tailrec` that would have the compiler error if it failed to TCO a function we expected it to be able to...

Reply:
Maybe I need to add inspection testing (http://arxiv.org/abs/1803.07130) to Motoko. I guess it’s a bit embarrassing that I didn’t already :-)

Reply:
We'll it doesn't overflow on large lists so I expect it's working... but I can double-check or just write the loop.